python/ray/tune/ray_trial_executor.py

```diff
@@ -906,6 +761,111 @@ def _change_working_directory(self, trial):
         else:
             yield
 
+    def get_next_future_result(self, next_trial_exists: bool) -> FutureResult:
+        """Get the next future result to be processed in TrialRunner.
+
+        This is a blocking wait with a timeout. The reason for the timeout is
+        we still want to print status info periodically in TrialRunner for
+        better user experience.
+
+        The handle of `RayTrialExecutor.STOP_RESULT` is purely internal to
+        RayTrialExecutor itself. All the other future results are handled by
+        TrialRunner.
+
+        In the future we may want to do most of the handle of
+        `RayTrialExecutor.RESTORE_RESULT` and `SAVING_RESULT` in
+        RayTrialExecutor itself and only notify TrialRunner to invoke
+        corresponding callbacks. This view is more consistent with our goal
+        of TrialRunner responsible for external facing Trial state transition,
+        while RayTrialExecutor responsible for internal facing transitions,
+        namely, `is_saving`, `is_restoring` etc.
+
+        Also you may notice that the boundary between RayTrialExecutor and
+        PlacementGroupManager right now is really blurred. This will be
+        improved once we move to an ActorPool abstraction.
+
+        `next_trial_exists` means that there is a trial to run - prioritize
+        returning PG_READY in this case. An example of `next_trial_exists`
+        being false is, e.g., synchronous hyperband, although there are pgs
+        ready, the scheduler is holding back scheduling new trials since the
+        whole band of trials is waiting for the slowest trial to finish. In
+        this case, we prioritize handling training result to avoid deadlock
+        situation.
```

Comment:
Let's make the prioritization crystal clear in the docstring and in the line comments. I think it's especially important to justify the prioritization:

1. If we have a new trial to run, we want to prioritize PG_READY futures so it can be scheduled as soon as possible
2. If a PG is already ready (e.g. though PG caching), we immediately return with PG_READY
3. Otherwise (if there is no new trial to run), we shuffle existing futures randomly so that we don't prioritize early trials over late trials in results processing.